### ISSUE TITLE
Detect dangling files

### ISSUE THREAD

**Issue by djmcgreal-cc** [remarkjs/remark-lint#308]
### Initial checklist

- [X] I read the [support docs](https://github.com/remarkjs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/remarkjs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/remarkjs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Aremarkjs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

Corpus of documentation. Using remark to detect invalid links and references. Would like to be able to detect dangling files, that is, files that are not reachable from a starting point.

### Solution

A rule that triggers against linted files are are absent from a directed acyclic graph of file links from a nominated starting file

### Alternatives

More elaborate processors have this sort of functionality, but I'm trying to keep the tooling here light and within the same ecosystem (i.e. remark/unified).
I've looked for something here, and for lightweight functionality in other tools but didn't find anything. If someone knows some, let me know!

**Comment by wooorm** [comment]
Hi!

This project here just checks one file and what’s inside it.

`remark-validate-links` can check across files, to see if a deep links to other files exist or not.
I doubt the tooling here would be very light if you want such functionality?
There are likely different clusters, that don’t connect to outside of them, but that do have references within them.

So, this is still a vague idea. I think you’d need to come up with examples of what input will give what output. And then how that would work in that project!

I also wonder about your use case, are you wondering about markdown? Or is this more about HTML files on some website? How many files? What types of content?

**Comment by djmcgreal-cc** [comment]
Thanks for your reply. 

I have nothing fancier than a git repository with markdown documents in them. They're checked for consistency in CI using, indeed, `remark-validate-links` (and `remark-lint-no-dead-urls` and a few others). Users are encouraged to start at the README.md and navigate outwards, there's not an overall index to pick up all *.md files in some hierarchy or otherwise convert to HTML, as I'm trying to keep the tooling as light as possible. Dangling files might be picked up by a search, or just browsing the repository, but otherwise they're not really integrated with the context of documentation flow and so I'd like to be able to check this as an aspect of consistency.

There may be clusters. I'd imagined the rule/plugin configuration would require a nominated entrypoint.

At the moment we're just using the CLI (also to keep tooling light, and because I know nothing about node), though I do plan to start using the API at some point so if I end up writing something, I'll let you know :)

**Comment by wooorm** [comment]
Assuming you use `remark-cli`, you could inject a tiny plugins:

```js
export default function myRemarkLoggingPlugin() {
    return function (_, file) {
        console.log(file.path, file.data)
    }
}
```

That should print metadata (at `remarkValidateLinksLandmarks` and `remarkValidateLinksReferences`). With that info, and an option from which file to treat as a root, and a bit of extra code, you should be able to connect whether files are referenced or not!

I’m open to discussing such functionality, but I’m currently, personally, not interested in taking the time to write it!
